### ISSUE TITLE
[Bug] $Shape<Field> can be assigned to ?Field

### ISSUE THREAD

**Issue by cprodescu** [facebook/flow#7730]
```
type Field1 = {
  a1: string,
  b1: string
};

function f(field1?: $Shape<Field1>) {
  const foo: ?Field1 = field1; // This is unsound. Not every `$Shape<Field1>` is a `?Field1`
}

type Field2 = {|
  a2: string,
  b2: string
|};

function f(field2?: $Shape<Field2>) {
  // Can also happen with exact fields, but only after a `null | undefined` check
  if (field2) {
    const foo: Field2 = field2; // This is unsound. Not every truthy `$Shape<Field2>` is a `Field2`
  }
}
```
[Repro](https://flow.org/try/#0C4TwDgpgBAYglhANgEwIxQLxQN4CgpQCGqAXFAM7ABOcAdgOYA0+UARqRdXfbgL4DcuXADMArrQDGwOAHtaUYQAphCFKgD8ZACQBlABaFIAHnhI0APgCUOFhLmUFMmWXWm1mBarT8oAel9QACp6cORQoVDi5DLiyAB0UAByMsBQEABuEFQgUAAGugbGbha54WGEea5eqLl8QqCQsF4ATB7YAD4shM1klDQMzASsPZz9PO0CQmKS0nIKyi2aUAWGECYtVjYE-lAAwoTyhIjRUIWQ8gDucMB6aQAehFKeZuSMbKKpcog5hMLAWUQ8rRRIhEFB2pFaMgICpaBBkKUJHoIBIANYsODCKALMzNax4AgEOy0BzCJxkYqtLAqXE+HbBCIRKIxKEJZKpDJZHLUD56HL5fSrda48ylCIVXKU2oEXh8IA)

Flow version: 0.98.1 (and lower, even 0.59.x)

## Expected behavior
Flow should print an error for the lines marked with `// This is unsound`, since not every `$Shape<Field1>` is a `?Field1` (e.g. `{ a1: '' }` is a `$Shape<Field1>`, but not a `?Field1`).

## Actual behavior
No errors are thrown

**Comment by noppa** [comment]
Duplicate of #7566, #7316, #5702, #6039, #6796.

AFAIK this is an unsound "feature" and not a bug. Or at least that's how it has always worked, the other issues haven't really gained much traction, and there's quite a bit of code that depends on its current behaviour.

A safe alternative to `$Shape` would be something like
```javascript
type Partial<T> = {...{...T}}
```

**Comment by goodmind** [comment]
Also `$Rest<T, {}>`

**Comment by goodmind** [comment]
Duplicate of #7316
